src/optimisation/Optimisation.jl

```diff
@@ -1,324 +1,327 @@
-
-Create a callable `OptimLogDensity` struct that evaluates a model using the given `context`.
-"""
-function OptimLogDensity(model::Model, context::OptimizationContext)
-    init = VarInfo(model)
-    return Turing.LogDensityFunction(init, model, context)
+struct ModeEstimationProblem{
+    M<:Model,
+    E<:ModeEstimator,
+    Iv<:AbstractVector,
+    Cons<:ModeEstimationConstraints,
+    OLD<:OptimLogDensity,
+    ADType<:AbstractADType,
+}
+    "The DynamicPPL model for which to estimate the mode."
+    model::M
+    """The mode estimator. Either `MLE()` for maximum likelihood or `MAP()` for maximum a
+    posteriori."""
+    estimator::E
+    "The initial value for the optimization."
+    init_value::Iv
+    "The constraints for the optimization problem."
+    constraints::Cons
+    "`OptimLogDensity` of the model, used as the objective function for the optimization."
+    log_density::OLD
+    "The automatic differentiation type to use. See ADTypes.jl."
+    adtype::ADType
+    "Whether the objective function is transformed to an unconstrained space."
+    linked::Bool
 end
 
 """
-    LogDensityProblems.logdensity(f::OptimLogDensity, z)
-
-Evaluate the negative log joint (with `DefaultContext`) or log likelihood (with `LikelihoodContext`)
-at the array `z`.
-"""
-function (f::OptimLogDensity)(z::AbstractVector)
-    varinfo = DynamicPPL.unflatten(f.varinfo, z)
-    return -getlogp(last(DynamicPPL.evaluate!!(f.model, varinfo, f.context)))
-end
-
-# NOTE: This seems a bit weird IMO since this is the _negative_ log-likelihood.
-LogDensityProblems.logdensity(f::OptimLogDensity, z::AbstractVector) = f(z)
+    ModeEstimationProblem(model, estimator, init_value, lb, ub, cons, lcons, ucons, adtype)
 
-function (f::OptimLogDensity)(F, G, z)
-    if G !== nothing
-        # Calculate negative log joint and its gradient.
-        # TODO: Make OptimLogDensity already an LogDensityProblems.ADgradient? Allow to specify AD?
-        ℓ = LogDensityProblemsAD.ADgradient(f)
-        neglogp, ∇neglogp = LogDensityProblems.logdensity_and_gradient(ℓ, z)
+Create a mode estimation problem.
 
-        # Save the gradient to the pre-allocated array.
-        copyto!(G, ∇neglogp)
-
-        # If F is something, the negative log joint is requested as well.
-        # We have already computed it as a by-product above and hence return it directly.
-        if F !== nothing
-            return neglogp
-        end
-    end
+`lb`, `ub`, `cons`, `lcons`, and `ucons` are constraints for the optimization problem as
+described in `ModeEstimationConstraints` and Optimization.jl.
 
-    # Only negative log joint requested but no gradient.
-    if F !== nothing
-        return LogDensityProblems.logdensity(f, z)
-    end
+If `init_value` is `nothing`, it is generated by either sampling from the prior distribution
+or uniformly from the box constraints.
 
-    return nothing
+The problem is always created with `linked=false`. See [`link`](@ref) for how to transform
+it.
```

Comment:
Is this needed? Is the linking not implicitly defined through the `varinfo`?

Reply:
Not needed anymore, see other comments.